core-rust/state-manager/src/accumulator_tree/storage.rs

```diff
@@ -0,0 +1,168 @@
+impl<T: ReadableAccuTreeStore<N> + WriteableAccuTreeStore<N>, N> AccuTreeStore<N> for T {}
+
+/// A vertical slice of an accumulator tree.
+/// This is an "incremental" persistence part of a tree representing a batch of appended leaf nodes
+/// and the resulting merkle updates, propagating up to a single root.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct TreeSlice<N> {
+    /// The tree-levels of this slice, arranged from leafs to root.
+    pub levels: Vec<TreeSliceLevel<N>>,
+}
+
+impl<N> TreeSlice<N> {
+    /// Creates a vertical slice directly from levels.
+    pub fn new(levels: Vec<TreeSliceLevel<N>>) -> Self {
+        Self { levels }
+    }
+
+    /// Returns the root of the tree.
+    pub fn root(&self) -> &N {
+        self.levels
+            .last()
+            .expect("empty slice")
+            .nodes
+            .first()
+            .expect("empty slice level")
+    }
+}
+
+/// A single tree-level of a `TreeSlice`.
+/// Effectively this means that it is a horizontal slice of a corresponding level of an entire
+/// accumulator tree.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct TreeSliceLevel<N> {
+    /// The cached left sibling of the first node of that level slice, i.e. present if and only if
+    /// the `nodes` actually start with a right child.
+    /// This value was not computed (i.e. not updated) during the append to the accumulator tree,
+    /// but it needed to be loaded from the storage (i.e. coming from the previous vertical slice).
+    /// It is stored in this slice again (i.e. cached), because it will be needed to produce merkle
+    /// proofs for the nodes contained in this slice.
+    pub left_sibling_cache: Option<N>,
+
+    /// The actual nodes computed during the append to the accumulator tree.
+    /// Depending on the level, these might be exactly all the appended leafs, or their composite
+    /// merkle nodes, all the way up. The highest level of each `TreeSlice` contains a single
+    /// element in the `nodes`, representing the merkle root.
+    pub nodes: Vec<N>,
+}
+
+impl<N> TreeSliceLevel<N> {
+    /// Creates a single horizontal level slice (of some vertical tree slice).
+    pub fn new(left_sibling_cache: Option<N>, nodes: Vec<N>) -> Self {
+        Self {
+            left_sibling_cache,
+            nodes,
+        }
+    }
+}
+
+/// An in-memory implementation of an `AccuTreeStore`.
+pub struct MemoryAccuTreeStore<N> {
+    /// Directly stored vertical slices, keyed by their end index (exclusive).
+    pub slices: HashMap<usize, TreeSlice<N>>,
+}
+
+impl<N> MemoryAccuTreeStore<N> {
+    /// Creates an empty in-memory storage.
+    pub fn new() -> Self {
+        Self {
+            slices: HashMap::new(),
+        }
+    }
+}
+
+impl<N: Clone> ReadableAccuTreeStore<N> for MemoryAccuTreeStore<N> {
+    fn get_tree_slice(&self, index: usize) -> Option<TreeSlice<N>> {
+        self.slices.get(&index).cloned()
```

Comment:
I see...
For "append to tree" and "generate proof" operations, this does not matter (because these operations are `>= O(M + logN)` already, probably with `C` far higher than that of `clone()` :P).
How about we add a dedicated O(1) `get_root_of_slice(index)` when needed?

(alternatively we can return a `&` from trait, but AFAIS our convention seems to be "return actual values from storage")

Reply:
I think reference might be more suitable, both here and generally. You can always clone afterwards.

I guess I'm more concerned about the `get_proof(index)` implementation (which can be implemented without the `M`).

Reply:
> I think reference might be more suitable, both here and generally.

That's an interesting view - but if I agree with it, then it actually follows that absolutely ALL apis (i.e. interfaces defined using `trait`s) in rust should only ever operate on references. But that's not what I observe.
I try to gauge what is the convention on "when to use ref vs val", and so far it looks like it is driven more by "conceptual" reasons (also "ownership" reasons?) than by pure performance. I mean, e.g. DB access is supposed to "produce a value" (out of some external source/representation), so it returns a val. _(in fact, our in-memory impl could only benefit from `&` because it happens to store the items directly [because it is only a test util/helper], while the real storage will be DB-backed, and returning a ref from there could be tricky)_